### PR TITLE
feat: Adding ksqlDB Query Status metric.

### PR DESCRIPTION
**Pull request by jnh5y**
### Description 

The existing metric `query-status` reports the state of the Kafka Streams application.

This new metric `ksql-query-status` reports the state of the ksqlDB.

This is being added to show that the query can be paused.  (While a ksqlDB query is paused,
the underlying Kafka Stream application is still in the `RUNNING` state.)

### Testing done 
Added unit tests and verified the new metric with JConsole locally.

### Reviewer checklist
- [ ] Ensure docs are updated if necessary. (eg. if a user visible feature is being added or changed).
- [ ] Ensure relevant issues are linked (description should include text like "Fixes #<issue number>")

### CODE REVIEW

**Comment by jnh5y** [comment]
I owe docs; wanted to get this up sooner to get feedback on the code changes.

File: docs/reference/metrics.md [hunk 1]
```diff
@@ -139,11 +139,20 @@ io.confluent.ksql.metrics:type=ksql-queries
 
 ### Attributes
 
+**ksqlDB query tatus**
+
+`ksql-query-status`
+
+The current ksqlDB status of the given query.   
+The metric `query-status` shows the Kafka Streams application state.
```
Comment (JimGalasyn) [inline docs/reference/metrics.md:147]:
```suggestion
The metric `query-status` shows the {{ site.kstreams }} application state.  
```

File: docs/reference/metrics.md [hunk 2]
```diff
@@ -139,11 +139,20 @@ io.confluent.ksql.metrics:type=ksql-queries
 
 ### Attributes
 
+**ksqlDB query tatus**
+
+`ksql-query-status`
+
+The current ksqlDB status of the given query.   
+The metric `query-status` shows the Kafka Streams application state.  
+The `PAUSE` / `RESUME` commands do not impact the Kafka Streams state, so this new metric shows when a query is paused.
```
Comment (JimGalasyn) [inline docs/reference/metrics.md:148]:
```suggestion
The `PAUSE` / `RESUME` commands do not impact the {{ site.kstreams }} state, so this new metric shows when a query is paused.
```

File: docs/reference/metrics.md [hunk 3]
```diff
@@ -139,11 +139,20 @@ io.confluent.ksql.metrics:type=ksql-queries
...
+
 **Query status**
 
 `query-status`
 
-The current status of the given query.
+The current Kafka Streams status of the given query.
```
Comment (JimGalasyn) [inline docs/reference/metrics.md:154]:
```suggestion
The current {{ site.kstreams }} status of the given query.  
```

File: docs/reference/metrics.md [hunk 4]
```diff
@@ -139,11 +139,20 @@ io.confluent.ksql.metrics:type=ksql-queries
...
+The metric `query-status` shows the Kafka Streams application state.  
+The `PAUSE` / `RESUME` commands do not impact the Kafka Streams state, so this new metric shows when a query is paused.
+
 **Query status**
 
 `query-status`
 
-The current status of the given query.
+The current Kafka Streams status of the given query.  
+Note that `ksql-query-status` has been added to show the ksqlDB query status.
```
Comment (JimGalasyn) [inline docs/reference/metrics.md:155]:
```suggestion
The `ksql-query-status` metric has been added to show the ksqlDB query status.
```

**Review by JimGalasyn — APPROVED** [review]
LGTM, with a few suggestions.

**Comment by vcrfxia** [comment]
Hey @jnh5y , this implementation departs from the pattern set by the existing implementation for the Kafka Streams state metric -- the Kafka Streams state metric has a listener mechanism to register changes in state, rather than simply calling `queryMetadata.getState()`. I'm not familiar with context for why the existing metric is set up the way that it is so I can't comment on whether or not something bad actually happens with this new pattern, but I do feel strongly that the two should follow the same pattern. If we want to migrate the existing metric onto the new pattern (after doing our due diligence to see if something bad is going to happen), that's fine too. Or we can switch the new metric to a listener as well since that's tried and true.

My best guess for why the existing metric is set up with a listener would be efficiency. If metrics are polled very frequently, it seems better to keep the state in the metric reporter rather than relying on the implementation for getting the state to be fast. Could be worth checking with @rodesai .

Also, while poking around the code for this PR, I realized I have no idea what KsqlQueryStatus is actually for, besides to represent whether a query is paused or not. Do you see either the ERROR or UNRESPONSIVE states actually being used anywhere? We have all sorts of logic for aggregating KsqlQueryStatus from multiple nodes, but I don't see anything that actually sets either the ERROR or UNRESPONSIVE state anywhere. I wonder if we have a cleanup opportunity here (in a separate PR).

**Comment by jnh5y** [comment]
Relative to `UNRESPONSIVE`, it is used in `ListQueriesExecutor`.  When query statuses are being calculated across the cluster, if a host is considered unresponsive, its queries are counted as unresponsive.

`ERROR` is used whenever the Kafka Streams application is in `KafkaStreams.State.ERROR`.

File: ksqldb-rest-app/src/test/java/io/confluent/ksql/rest/integration/PauseResumeIntegrationTest.java [hunk 5]
```diff
@@ -125,6 +127,13 @@ public void shouldPauseAndResumeQuery() {
     RestIntegrationTestUtil.makeKsqlRequest(REST_APP, "PAUSE " + queryId + ";");
 
     assertThat(getPausedCount(), equalTo(1L));
+    REST_APP.getEngine().metricCollectors().getMetrics();
```
Comment (vcrfxia) [inline ksqldb-rest-app/src/test/java/io/confluent/ksql/rest/integration/PauseResumeIntegrationTest.java:130]: Is this call necessary? The return value is ignored.

File: ksqldb-rest-app/src/test/java/io/confluent/ksql/rest/integration/PauseResumeIntegrationTest.java [hunk 6]
```diff
@@ -125,6 +127,13 @@ public void shouldPauseAndResumeQuery() {
...
+    final MetricName key = REST_APP.getEngine().metricCollectors().getMetrics().metrics().keySet()
+            .stream()
+            .filter(metricName -> metricName.name().contains("ksql-query-status"))
+            .findFirst().get();
+    final KafkaMetric metric = REST_APP.getEngine().metricCollectors().getMetrics().metric(key);
+    assertThat((String)metric.metricValue(), equalTo("PAUSED"));
```
Comment (vcrfxia) [inline ksqldb-rest-app/src/test/java/io/confluent/ksql/rest/integration/PauseResumeIntegrationTest.java:136]: Also check that the metric is as expected after we RESUME?

File: ksqldb-engine/src/main/java/io/confluent/ksql/engine/QueryEventListener.java [hunk 7]
```diff
@@ -54,6 +54,15 @@ default void onStateChange(
   ) {
   }
 
+  /**
+   * Called when the state of the ksqlDB query metadata changes
```
Comment (vcrfxia) [inline ksqldb-engine/src/main/java/io/confluent/ksql/engine/QueryEventListener.java:58]: This comment makes it sound like the method is called anytime any part of the ksqlDB query metadata changes, but it's only called when the query is paused or resumed, right? If so, let's update the comment to reflect that.

**Review by vcrfxia — APPROVED** [review]
Thanks @jnh5y -- LGTM besides the comments inline.

File: ksqldb-rest-app/src/test/java/io/confluent/ksql/rest/integration/PauseResumeIntegrationTest.java (hunk 5 above)
Comment (jnh5y) [reply ksqldb-rest-app/src/test/java/io/confluent/ksql/rest/integration/PauseResumeIntegrationTest.java:130]: Nope, it is going.

File: ksqldb-rest-app/src/test/java/io/confluent/ksql/rest/integration/PauseResumeIntegrationTest.java (hunk 6 above)
Comment (jnh5y) [reply ksqldb-rest-app/src/test/java/io/confluent/ksql/rest/integration/PauseResumeIntegrationTest.java:136]: Good idea; doing it.

File: ksqldb-engine/src/main/java/io/confluent/ksql/engine/QueryEventListener.java (hunk 7 above)
Comment (jnh5y) [reply ksqldb-engine/src/main/java/io/confluent/ksql/engine/QueryEventListener.java:58]: Presently that's when it is called.  Updating the java doc.

**Pull request merged**